File: src/main/resources/vertx-mqtt-server-js/mqtt_will.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-mqtt-server-js/mqtt_will */
var utils = require('vertx-js/util/utils');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JMqttWill = Java.type('io.vertx.mqtt.MqttWill');

/**
 Will information from the remote MQTT client

 @class
*/
var MqttWill = function(j_val) {

  var j_mqttWill = j_val;
  var that = this;

  /**

   @public

   @return {boolean} Will flag for indicating the will message presence
   */
  this.isWillFlag = function() {
    var __args = arguments;
    if (__args.length === 0) {
      if (that.cachedisWillFlag == null) {
        that.cachedisWillFlag = j_mqttWill["isWillFlag()"]();
      }
      return that.cachedisWillFlag;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {string} Topic for the will as provided by the remote MQTT client
   */
  this.willTopic = function() {
    var __args = arguments;
    if (__args.length === 0) {
      if (that.cachedwillTopic == null) {
        that.cachedwillTopic = j_mqttWill["willTopic()"]();
      }
      return that.cachedwillTopic;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {string} Payload for the will as provided by the remote MQTT client
   */
  this.willMessage = function() {
    var __args = arguments;
    if (__args.length === 0) {
      if (that.cachedwillMessage == null) {
        that.cachedwillMessage = j_mqttWill["willMessage()"]();
      }
      return that.cachedwillMessage;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {number} QoS level for the will as provided by the remote MQTT client
   */
  this.willQos = function() {
    var __args = arguments;
    if (__args.length === 0) {
      if (that.cachedwillQos == null) {
        that.cachedwillQos = j_mqttWill["willQos()"]();
      }
      return that.cachedwillQos;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {boolean} If the will must be retained as provided by the remote MQTT client
   */
  this.isWillRetain = function() {
    var __args = arguments;
    if (__args.length === 0) {
      if (that.cachedisWillRetain == null) {
        that.cachedisWillRetain = j_mqttWill["isWillRetain()"]();
      }
      return that.cachedisWillRetain;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_mqttWill;
};

MqttWill._jclass = utils.getJavaClass("io.vertx.mqtt.MqttWill");
MqttWill._jtype = {
  accept: function(obj) {
    return MqttWill._jclass.isInstance(obj._jdel);
  },
  wrap: function(jdel) {
    var obj = Object.create(MqttWill.prototype, {});
    MqttWill.apply(obj, arguments);
    return obj;
  },
  unwrap: function(obj) {
    return obj._jdel;
  }
};
MqttWill._create = function(jdel) {
  var obj = Object.create(MqttWill.prototype, {});
  MqttWill.apply(obj, arguments);
  return obj;
}
module.exports = MqttWill;

File: src/main/java/io/vertx/mqtt/MqttWill.java
/*
 * Copyright 2016 Red Hat Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.vertx.mqtt;

import io.netty.handler.codec.mqtt.MqttProperties;
import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import static io.netty.handler.codec.mqtt.MqttProperties.MqttPropertyType.*;
/**
 * Will information from the remote MQTT client
 */
@DataObject
public class MqttWill {

  private final boolean isWillFlag;
  private final String willTopic;
  private final Buffer willMessage;
  private final int willQos;
  private final boolean isWillRetain;
  private final MqttProperties willProperties;

  /**
   * Constructor
   *
   * @param isWillFlag   indicates will message presence
   * @param willTopic    topic to publish the will
   * @param willMessage  payload of the will
   * @param willQos      qos level for the will
   * @param isWillRetain if the will message must be retained
   * @param willProperties MQTT properties of the last will message
   */
  public MqttWill(boolean isWillFlag, String willTopic, Buffer willMessage, int willQos, boolean isWillRetain, MqttProperties willProperties) {
    this.isWillFlag = isWillFlag;
    this.willTopic = willTopic;
    this.willMessage = willMessage;
    this.willQos = willQos;
    this.isWillRetain = isWillRetain;
    this.willProperties = willProperties;
  }

  /**
   * Create instance from JSON
   *
   * @param json the JSON
   */
  public MqttWill(JsonObject json) {
    this.isWillFlag = json.getBoolean("isWillFlag");
    this.willTopic = json.getString("willTopic");
    this.willMessage = json.getBuffer("willMessage");
    this.willQos = json.getInteger("willQos");
    this.isWillRetain = json.getBoolean("isWillRetain");
    this.willProperties = propertiesFromJson(json.getJsonArray("willProperties"));
  }

  /**
   * @return the will flag for indicating the will message presence
   */
  public boolean isWillFlag() {
    return this.isWillFlag;
  }

  /**
   * @return the topic for the will as provided by the remote MQTT client
   */
  public String getWillTopic() {
    return this.willTopic;
  }

  /**
   * @return the payload for the will as provided by the remote MQTT client
   */
  public Buffer getWillMessage() {
    return this.willMessage;
  }

  /**
   * @return the payload for the will as provided by the remote MQTT client
   */
  public byte[] getWillMessageBytes() {
    return this.willMessage != null ? this.willMessage.getBytes() : null;
  }

  /**
   * @return the QoS level for the will as provided by the remote MQTT client
   */
  public int getWillQos() {
    return this.willQos;
  }

  /**
   * @return true if the will must be retained as provided by the remote MQTT client
   */
  public boolean isWillRetain() {
    return this.isWillRetain;
  }

  /**
   * @return MQTT properties of the last will message
   */
  public MqttProperties getWillProperties()  {
    return this.willProperties;
  }

  /**
   * Convert instance in JSON
   *
   * @return JSON representation
   */
  public JsonObject toJson() {
    JsonObject json = new JsonObject();
    json.put("isWillFlag", this.isWillFlag);
    json.put("willTopic", this.willTopic);
    json.put("willMessage", this.willMessage);
    json.put("willQos", this.willQos);
    json.put("isWillRetain", this.isWillRetain);
    json.put("willProperties", propertiesToJson(this.willProperties));
    return json;
  }

  public static JsonArray propertiesToJson(MqttProperties properties) {
    JsonArray array = new JsonArray();
    for(MqttProperties.MqttProperty<?> prop: properties.listAll()) {
      array.add(propertyToJson(prop));
    }
    return array;
  }

  public static JsonObject propertyToJson(MqttProperties.MqttProperty<?> prop) {
    JsonObject obj = new JsonObject();
    if(prop instanceof MqttProperties.StringProperty ||
      prop instanceof MqttProperties.IntegerProperty) {
      obj.put("id", prop.propertyId());
      obj.put("val", prop.value());
    } else if(prop instanceof MqttProperties.BinaryProperty) {
      obj.put("id", prop.propertyId());
      String value = Base64.getEncoder().encodeToString(((MqttProperties.BinaryProperty) prop).value());
      obj.put("val", value);
    } else if(prop instanceof MqttProperties.UserProperties) {
      for(MqttProperties.StringPair kv: ((MqttProperties.UserProperties) prop).value()) {
        obj.put("id", prop.propertyId());
        obj.put("key", kv.key);
        obj.put("val", kv.value);
      }
    }
    return obj;
  }

  public static MqttProperties propertiesFromJson(JsonArray array) {
    MqttProperties props = new MqttProperties();
    for(Object item: array) {
      props.add(propertyFromJson((JsonObject) item));
    }
    return props;
  }

  public static MqttProperties.MqttProperty<?> propertyFromJson(JsonObject obj) {
    int id = obj.getInteger("id");

    MqttProperties.MqttPropertyType propType = MqttProperties.MqttPropertyType.valueOf(id);
    switch (propType) {
      case PAYLOAD_FORMAT_INDICATOR:
      case REQUEST_PROBLEM_INFORMATION:
      case REQUEST_RESPONSE_INFORMATION:
      case MAXIMUM_QOS:
      case RETAIN_AVAILABLE:
      case WILDCARD_SUBSCRIPTION_AVAILABLE:
      case SUBSCRIPTION_IDENTIFIER_AVAILABLE:
      case SHARED_SUBSCRIPTION_AVAILABLE:
      case SERVER_KEEP_ALIVE:
      case RECEIVE_MAXIMUM:
      case TOPIC_ALIAS_MAXIMUM:
      case TOPIC_ALIAS:
      case PUBLICATION_EXPIRY_INTERVAL:
      case SESSION_EXPIRY_INTERVAL:
      case WILL_DELAY_INTERVAL:
      case MAXIMUM_PACKET_SIZE:
      case SUBSCRIPTION_IDENTIFIER:
        return new MqttProperties.IntegerProperty(id, obj.getInteger("val"));
      case CONTENT_TYPE:
      case RESPONSE_TOPIC:
      case ASSIGNED_CLIENT_IDENTIFIER:
      case AUTHENTICATION_METHOD:
      case RESPONSE_INFORMATION:
      case SERVER_REFERENCE:
      case REASON_STRING:
        return new MqttProperties.StringProperty(id, obj.getString("val"));
      case CORRELATION_DATA:
      case AUTHENTICATION_DATA:
        return new MqttProperties.BinaryProperty(id, Base64.getDecoder().decode(obj.getString("val")));
      case USER_PROPERTY:
        String key = obj.getString("key");
        return new MqttProperties.UserProperty(key, obj.getString("val"));
      default:
        throw new IllegalArgumentException("Unsupported property type: " + propType);
    }
  }
}
